libopae/api-shell.c

```diff
@@ -304,19 +346,34 @@ fpga_result fpgaUpdateProperties(fpga_token token, fpga_properties prop)
 
 	// If the output properties has a parent token set,
 	// then it will be a raw token. We need to wrap it.
-	if (fpgaPropertiesGetParent(prop, &parent) == FPGA_OK) {
-		opae_wrapped_token *wrapped_parent =
-			opae_allocate_wrapped_token(
-				parent, wrapped_token->adapter_table);
 
-		if (!wrapped_parent) {
-			OPAE_ERR("malloc failed");
-			return FPGA_NO_MEMORY;
-		}
+	p = opae_validate_and_lock_properties(prop);
+
+	ASSERT_NOT_NULL(p);
 
-		return fpgaPropertiesSetParent(prop, wrapped_parent);
+	if (FIELD_VALID(p, FPGA_PROPERTY_PARENT)) {
+		fpga_token parent = NULL;
+
+		res = wrapped_token->adapter_table->fpgaCloneToken(
+			p->parent, &parent);
+
+		if (res == FPGA_OK) {
+			opae_wrapped_token *wrapped_parent =
+				opae_allocate_wrapped_token(
+					parent, wrapped_token->adapter_table);
+
+			if (wrapped_parent) {
+				p->parent = wrapped_parent;
+			} else {
+				OPAE_ERR("malloc failed");
+				res = FPGA_NO_MEMORY;
+			}
+
+		}
```

Comment:
Same thing here about refactoring into a common function

Reply:
Let's wait and see what pans out on the clone discussion.